query-grammar/src/query_grammar.rs

```diff
@@ -374,6 +374,25 @@ pub fn parse_to_ast<'a>() -> impl Parser<&'a str, Output = UserInputAst> {
     spaces()
         .with(optional(ast()).skip(eof()))
         .map(|opt_ast| opt_ast.unwrap_or_else(UserInputAst::empty_query))
+        .map(|ast| rewrite_ast(ast))
+}
+
+/// Removes unnecessary children clauses in AST
+///
+/// Motivated by [issue #1433](https://github.com/quickwit-oss/tantivy/issues/1433)
+fn rewrite_ast(input: UserInputAst) -> UserInputAst {
+    if let UserInputAst::Clause(terms) = input {
+        UserInputAst::Clause(terms.into_iter().map(rewrite_ast_clause).collect())
+    } else {
+        input
+    }
+}
+
+fn rewrite_ast_clause(input: (Option<Occur>, UserInputAst)) -> (Option<Occur>, UserInputAst) {
+    match input {
+        (None, UserInputAst::Clause(mut clauses)) if clauses.len() == 1 => clauses.remove(0),
```

Comment:
I fear that the compiler might not be able to remove that part of the `Vec::remove`'s code which does not apply for a vector of length one. Maybe this could use `clauses.pop().unwrap()` to guide it towards a simpler method to begin with?

Reply:
Yes, indeed. Forgot about `pop()`. Thank you.